uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php

```diff
@@ -138,6 +147,232 @@ private static function validateMandatory($params) {
         'to_date_type'
       );
     }
+
+    if (empty($params['request_type'])) {
+      throw new InvalidLeaveRequestException(
+        'The request_type should not be empty',
+        'leave_request_empty_request_type',
+        'request_type'
+      );
+    }
+  }
+
+  /**
+   * Validates the values of the request_type field
+   *
+   * @param array $params
+   *
+   * @throws \CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   */
+  private static function validateRequestType($params) {
+    $validRequestTypes = [
+      self::REQUEST_TYPE_LEAVE,
+      self::REQUEST_TYPE_SICKNESS,
+      self::REQUEST_TYPE_TOIL,
+      self::REQUEST_TYPE_PUBLIC_HOLIDAY
+    ];
+
+    if(!in_array($params['request_type'], $validRequestTypes)) {
+      throw new InvalidLeaveRequestException(
+        'The request_type is invalid',
+        'leave_request_invalid_request_type',
+        'request_type'
+      );
+    }
+  }
+
+  /**
+   * Validates the TOIL mandatory fields according to the value of request_type.
+   *
+   * @param array $params
+   *
+   * @throws \CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   */
+  private static function validateTOILMandatoryFields($params) {
+    $toilRequiredFields = [ 'toil_duration', 'toil_to_accrue' ];
+    $toilFields = array_merge($toilRequiredFields, [ 'toil_expiry_date' ]);
+
+    if(self::REQUEST_TYPE_TOIL == $params['request_type']) {
+      foreach($toilRequiredFields as $requiredField) {
```

Comment:
It would greatly reduce/improve the code in LeaveRequest::validateMandatory If we use this same approach for validating mandatory fields for a leave request. We could just have a `$leaveRequestRequiredFields` array and loop through .